### PR TITLE
Handle reference error types in `ensure_started` and `let_error`

### PR DESCRIPTION
**Pull request by aurianer**
Fixes partially #284

### CODE REVIEW

**Comment by aurianer** [comment]
bors try

**Comment by aurianer** [comment]
bors try

File: libs/pika/execution/include/pika/execution/algorithms/ensure_started.hpp [hunk 1]
```diff
@@ -115,8 +115,10 @@ namespace pika::execution::experimental {
             template <template <typename...> class Variant>
             using error_types =
                 pika::util::detail::unique_t<pika::util::detail::prepend_t<
-                    typename pika::execution::experimental::sender_traits<
-                        Sender>::template error_types<Variant>,
+                    pika::util::detail::transform_t<
```
Comment (msimberg) [inline libs/pika/execution/include/pika/execution/algorithms/ensure_started.hpp:118]: Could you update `value_types` as well (with a corresponding test)?

**Review by msimberg — CHANGES_REQUESTED** [review]
Looks good, but I still have to have a closer look at `let_value/error` again.

Could you update the PR title to mention the updated algorithms?

File: libs/pika/execution/include/pika/execution/algorithms/ensure_started.hpp (hunk 1 above)
Comment (aurianer) [reply libs/pika/execution/include/pika/execution/algorithms/ensure_started.hpp:118]: As discussed, already tested here:
https://github.com/pika-org/pika/blob/5342bce01424dc1fcb78dec0578861b3a3fe7354/libs/pika/execution/tests/unit/algorithm_ensure_started.cpp#L104

**Comment by aurianer** [comment]
bors try

File: libs/pika/execution/tests/unit/algorithm_then.cpp [hunk 2]
```diff
@@ -75,6 +75,18 @@ int main()
         PIKA_TEST(set_value_called);
     }
 
+    {
+        std::atomic<bool> set_value_called{false};
+        int x = 42;
+        auto s = ex::then(const_reference_sender<decltype(x)>{x},
+            [](int i) { return i + 1; });    // generic lambda
```
Comment (msimberg) [inline libs/pika/execution/tests/unit/algorithm_then.cpp:82]: I think the trailing comment doesn't apply anymore and can be removed.

**Review by msimberg — CHANGES_REQUESTED** [review]
Except for the one unnecessary comment this looks good.

**Comment by aurianer** [comment]
bors merge

**Review by msimberg — APPROVED** [review]

**Comment by msimberg** [comment]
bors merge

**Comment by msimberg** [comment]
bors merge

**Pull request merged**